File: rmp/frontend/pages/pool/[id]/squares.js
import { useEffect, useMemo, useState } from 'react';
import { useRouter } from 'next/router';
import Link from 'next/link';
import ProtectedRoute from '../../../components/ProtectedRoute';
import { PoolWorkspaceNav, WorkspaceHeader } from '../../../components/ProductWorkspace';
import { useAuth } from '../../../context/AuthContext';

const authHeaders = () => ({ Authorization: `Bearer ${localStorage.getItem('access_token')}` });
const dollarsFromCents = (cents) => cents == null ? '' : (cents / 100).toFixed(2);

export default function SquaresPage() {
  const { query } = useRouter();
  const { user } = useAuth();
  const [board, setBoard] = useState(null);
  const [error, setError] = useState('');
  const [busy, setBusy] = useState(false);
  const [pot, setPot] = useState('');
  const [potMode, setPotMode] = useState('fixed');
  const [pendingSquare, setPendingSquare] = useState(null);
  const [claimName, setClaimName] = useState('');
  const claims = useMemo(() => Object.fromEntries((board?.claims || []).map((c) => [`${c.row_index}-${c.column_index}`, c])), [board]);
  const reservations = useMemo(() => {
    const grouped = new Map();
    for (const claim of board?.claims || []) {
      const displayName = claim.display_name || 'Reserved player';
      const key = `${claim.user_id}:${displayName}`;
      const reservation = grouped.get(key) || { key, claim_id: claim.id, user_id: claim.user_id, display_name: displayName, count: 0, blocks: [] };
      reservation.count += 1;
      reservation.blocks.push(claim.block_number ?? (claim.row_index * 10 + claim.column_index + 1));
      grouped.set(key, reservation);
    }
    return Array.from(grouped.values()).map((reservation) => ({ ...reservation, blocks: reservation.blocks.sort((a, b) => a - b) })).sort((a, b) => a.display_name.localeCompare(b.display_name));
  }, [board]);

  const load = async () => {
    const response = await fetch(`${process.env.NEXT_PUBLIC_API_URL}/squares/${query.id}`, { headers: authHeaders() });
    const data = await response.json().catch(() => ({}));
    if (!response.ok) throw new Error(data.detail || 'Unable to load the Squares board.');
    const mode = data.pot_mode || 'fixed';
    setBoard(data); setPotMode(mode); setPot(dollarsFromCents(mode === 'per_square' ? data.per_square_cents : data.total_pot_cents));
  };
  useEffect(() => { if (query.id) load().catch((err) => setError(err.message)); }, [query.id]);

  const choose = async (row, column) => {
    const claim = claims[`${row}-${column}`];
    setPendingSquare({ row, column, blockNumber: row * 10 + column + 1, claim });
    setClaimName(claim?.display_name || '');
    setError('');
  };

  const saveSquare = async (event) => {
    event.preventDefault();
    const normalizedDisplayName = claimName.trim().replace(/\s+/g, ' ');
    if (!pendingSquare.claim && !normalizedDisplayName) return;
    setBusy(true); setError('');
    try {
      const editing = Boolean(pendingSquare.claim);
      const response = await fetch(`${process.env.NEXT_PUBLIC_API_URL}/squares/${query.id}/claims${editing ? '/display-name' : ''}`, {
        method: editing ? 'PATCH' : 'POST', headers: { ...authHeaders(), 'Content-Type': 'application/json' },
        body: JSON.stringify(editing
          ? { claim_id: pendingSquare.claim.id, display_name: normalizedDisplayName }
          : { row_index: pendingSquare.row, column_index: pendingSquare.column, display_name: normalizedDisplayName }),
      });
      const data = await response.json().catch(() => ({}));
      if (!response.ok) throw new Error(data.detail || 'Unable to update that square.');
      setPendingSquare(null); setClaimName('');
      if (editing) setBoard(data); else await load();
    } catch (err) { setError(err.message); } finally { setBusy(false); }
  };

  const lock = async () => {
    if (!confirm('Lock this board and permanently randomize both score axes?')) return;
    setBusy(true); setError('');
    try {
      const response = await fetch(`${process.env.NEXT_PUBLIC_API_URL}/squares/${query.id}/lock`, { method: 'POST', headers: authHeaders() });
      const data = await response.json().catch(() => ({}));
      if (!response.ok) throw new Error(data.detail || 'Unable to lock the board.');
      setBoard(data);
    } catch (err) { setError(err.message); } finally { setBusy(false); }
  };

  const savePot = async (event) => {
    event.preventDefault(); setBusy(true); setError('');
    try {
      const response = await fetch(`${process.env.NEXT_PUBLIC_API_URL}/squares/${query.id}/payouts`, {
        method: 'PATCH', headers: { ...authHeaders(), 'Content-Type': 'application/json' },
        body: JSON.stringify({
          pot_mode: potMode,
          total_pot_cents: potMode === 'fixed' && pot !== '' ? Math.round(Number(pot) * 100) : null,
          per_square_cents: potMode === 'per_square' && pot !== '' ? Math.round(Number(pot) * 100) : null,
          q1_percent: 25, halftime_percent: 25, q3_percent: 25, final_percent: 25,
        }),
      });
      const data = await response.json().catch(() => ({}));
      if (!response.ok) throw new Error(data.detail || 'Unable to save payout settings.');
      setBoard(data);
    } catch (err) { setError(err.message); } finally { setBusy(false); }
  };

  if (!board) return <ProtectedRoute><main className="product-page-shell"><p>{error || 'Loading Squares board…'}</p></main></ProtectedRoute>;
  const games = board.games?.length ? board.games : [board.game];
  const game = games[0];
  const gameTitle = games.length === 1 ? `${game.away_team.abbrv} at ${game.home_team.abbrv}` : `${games.length} selected games`;
  return <ProtectedRoute><main className="product-page-shell squares-page">
    <PoolWorkspaceNav poolId={query.id} poolName={board.pool_name} poolType="squares" active="entries" showAdmin={board.permissions.is_admin} />
    <WorkspaceHeader eyebrow={board.locked ? 'Board locked' : board.plan === 'free' ? 'Owner-managed board' : 'Choose your squares'} title={gameTitle} description={`Locks ${new Date(board.lock_time || game.start_time).toLocaleString()} · ${board.claims.length}/${board.block_limit ?? 100} blocks reserved`} actions={<div className="squares-screen-actions"><button type="button" onClick={() => window.print()}>Print / Save PDF</button>{board.permissions.is_admin && !board.locked && <button type="button" onClick={lock} disabled={busy}>Lock & randomize digits</button>}</div>} />
    <header className="squares-print-header">
      <span>Run My Pool · Squares</span>
      <h1>{board.pool_name}</h1>
      <p>{games.map((item) => `${item.away_team.abbrv} at ${item.home_team.abbrv}`).join(' · ')}</p>
    </header>
    {error && <div className="workspace-alert workspace-alert--error">{error}</div>}
    {board.permissions.is_admin && board.plan === 'free' && <aside className="squares-upgrade"><div><span>Free · Owner-managed</span><strong>All 100 blocks are available for you to manage</strong><p>Collect selections from players yourself, enter their display name, and reserve their requested blocks. Players cannot join this free board online. Upgrade to Squares Plus to invite players and let them reserve their own blocks.</p></div><Link href="/pricing?checkout=squares-plus">Enable online player access for $10</Link></aside>}
    {board.permissions.is_admin && board.plan === 'squares-plus' && <aside className="squares-upgrade"><div><span>Squares Plus</span><strong>All 100 self-service blocks are open</strong><p>Upgrade to Commish for admin member assignment and per-reservation pots.</p></div><Link href="/pricing?checkout=commissioner">Upgrade to Commish for $29</Link></aside>}
    <section className="squares-summary" aria-label="Squares pool summary">
      <div><span>Total pot</span><strong>{board.total_pot_cents == null ? 'Not set' : `$${(board.total_pot_cents / 100).toFixed(2)}`}</strong>{board.pot_mode === 'per_square' && board.per_square_cents != null && <small>${(board.per_square_cents / 100).toFixed(2)} per reserved block</small>}</div>
      <div><span>Reserved</span><strong>{board.claims.length}/{board.block_limit ?? 100} included</strong><small>100 blocks on the board</small></div>
      <div><span>Score digits</span><strong>{board.locked ? 'Randomized' : 'Hidden until lock'}</strong></div>
    </section>
    <p className="squares-instructions">Home score runs down the left. Away score runs across the top. {!board.locked && 'Score digits are generated randomly by the server when the board locks.'}</p>
    <section className="squares-games" aria-labelledby="squares-games-heading"><div><h2 id="squares-games-heading">Games on this board</h2><p>The same grid and score digits apply to every matchup.</p></div><ul>{games.map((item) => <li key={item.game_id}><strong>{item.away_team.abbrv} at {item.home_team.abbrv}</strong><span>{new Date(item.start_time).toLocaleString()}</span></li>)}</ul></section>
    <div className="squares-scroll"><div className="squares-grid" role="grid" aria-label="10 by 10 Squares board">
      <div className="squares-corner">HOME ↓<br />AWAY →</div>
      {Array.from({ length: 10 }, (_, col) => <div className="squares-axis" key={`a${col}`}>{board.away_digits?.[col] ?? '?'}</div>)}
      {Array.from({ length: 10 }, (_, row) => <div className="squares-row" key={row}>
        <div className="squares-axis">{board.home_digits?.[row] ?? '?'}</div>
        {Array.from({ length: 10 }, (_, col) => { const claim = claims[`${row}-${col}`]; const blockNumber = row * 10 + col + 1; const mayRelease = claim?.user_id === user?.id || board.permissions.is_admin; const owner = claim?.display_name || 'Reserved player'; return <button key={col} role="gridcell" aria-label={claim ? `Block ${blockNumber}, reserved by ${owner}` : `Block ${blockNumber}, available`} disabled={busy || board.locked || (!board.permissions.can_claim && !claim) || (claim && !mayRelease)} className={`squares-cell ${claim ? 'is-claimed' : ''} ${claim?.user_id === user?.id ? 'is-mine' : ''}`} onClick={() => choose(row, col)} title={claim ? `Block ${blockNumber} · Reserved by ${owner}` : `Block ${blockNumber} · Available`}>{claim ? <span className="squares-block-owner">{owner}</span> : <span className="squares-block-number">{blockNumber}</span>}</button>; })}
      </div>)}
    </div></div>
    {pendingSquare && <div className="squares-claim-backdrop" onMouseDown={(event) => event.target === event.currentTarget && !busy && setPendingSquare(null)}><section className="squares-claim-dialog" role="dialog" aria-modal="true" aria-labelledby="squares-claim-title"><form onSubmit={saveSquare}><header><span>Block {pendingSquare.blockNumber}</span><h2 id="squares-claim-title">{pendingSquare.claim ? 'Edit this block' : 'Name this block'}</h2><p>{pendingSquare.claim ? 'Change the display name, or clear it to unclaim this square.' : 'Enter the name that should appear inside this square.'}</p></header><label>Block name<input type="text" maxLength="100" autoComplete="nickname" value={claimName} onChange={(event) => setClaimName(event.target.value)} placeholder="Example: Tony S." autoFocus /></label><footer><button type="button" onClick={() => { setPendingSquare(null); setClaimName(''); }} disabled={busy}>Cancel</button><button type="submit" disabled={busy || (!pendingSquare.claim && !claimName.trim())}>{pendingSquare.claim ? (claimName.trim() ? 'Save block' : 'Unclaim block') : 'Reserve block'}</button></footer></form></section></div>}
    <section className="squares-reservations" aria-labelledby="squares-reservations-heading">
      <div><h2 id="squares-reservations-heading">Reservations</h2><p>Everyone in the pool can see who has reserved squares.</p></div>
      {reservations.length ? <ul>{reservations.map((reservation) => <li key={reservation.key}><div><strong>{reservation.display_name}</strong><span>{reservation.count} {reservation.count === 1 ? 'square' : 'squares'} · Blocks {reservation.blocks.join(', ')}</span></div></li>)}</ul> : <p>No squares have been reserved yet.</p>}
    </section>
    <section className="squares-results"><h2>Quarter winners</h2>{games.map((selectedGame) => <div className="squares-game-results" key={selectedGame.game_id}><h3>{selectedGame.away_team.abbrv} at {selectedGame.home_team.abbrv} results</h3><div>{['q1', 'halftime', 'q3', 'final'].map((name) => { const result = board.payouts.find((item) => (item.game_id == null || item.game_id === selectedGame.game_id) && item.checkpoint === name); return <article key={name}><strong>{name === 'q1' ? '1st Quarter' : name === 'q3' ? '3rd Quarter' : name[0].toUpperCase() + name.slice(1)}</strong>{result ? <><span>{result.away_score}–{result.home_score}</span><b>{result.winner_display_name || (result.winner_email ? 'Reserved player' : 'Unclaimed square')}</b>{result.amount_cents != null && <small>${(result.amount_cents / 100).toFixed(2)} recorded</small>}</> : <span>Pending</span>}</article>; })}</div></div>)}</section>
    {board.permissions.is_admin && !board.locked && <form className="squares-pot" onSubmit={savePot}><div><h2>Admin payout setup</h2><p>Choose one fixed pot or let the total grow with every reserved block. Quarter payouts remain 25% each. Run My Pool records winners and amounts but does not move money.</p><fieldset><legend>Pot calculation</legend><label className={potMode === 'fixed' ? 'is-selected' : ''}><input type="radio" name="pot-mode" value="fixed" checked={potMode === 'fixed'} onChange={() => { setPotMode('fixed'); setPot(dollarsFromCents(board.pot_mode === 'fixed' ? board.total_pot_cents : null)); }} /> Fixed total</label><label className={potMode === 'per_square' ? 'is-selected' : ''} aria-disabled={!board.permissions.can_use_variable_pot}><input type="radio" name="pot-mode" value="per_square" checked={potMode === 'per_square'} disabled={!board.permissions.can_use_variable_pot} onChange={() => { setPotMode('per_square'); setPot(dollarsFromCents(board.per_square_cents)); }} /> Per reserved block {!board.permissions.can_use_variable_pot && <small>Commish</small>}</label></fieldset></div><label>{potMode === 'fixed' ? 'Total pot ($)' : 'Amount per reserved block ($)'}<input type="number" min="0" step="0.01" inputMode="decimal" value={pot} onChange={(event) => setPot(event.target.value)} placeholder={potMode === 'fixed' ? 'Optional' : 'Required'} required={potMode === 'per_square'} /></label><button disabled={busy}>Save payouts</button></form>}
  </main></ProtectedRoute>;
}

File: rmp/backend/squares.py
"""Authoritative single-game 10x10 Squares boards."""

import json
import secrets
import uuid
from datetime import datetime, timezone

from fastapi import APIRouter, Depends, HTTPException, Response, status
from sqlalchemy.exc import IntegrityError
from sqlalchemy.orm import Session, joinedload

import deps
import entitlements
import models
import schemas
from audit_utils import create_audit_log
from platform_admin import is_platform_super_admin
from pool_access import is_pool_participant

router = APIRouter(prefix="/squares", tags=["squares"])


def _now() -> datetime:
    return datetime.now(timezone.utc).replace(tzinfo=None)


def _pool(db: Session, pool_id: str) -> models.Pool:
    pool = (
        db.query(models.Pool)
        .options(joinedload(models.Pool.square_board))
        .filter(models.Pool.id == pool_id)
        .first()
    )
    if not pool or pool.pool_type != "squares" or not pool.square_board:
        raise HTTPException(status_code=404, detail="Squares pool not found")
    return pool


def _selected_games(pool: models.Pool) -> list[models.Schedule]:
    games = [selection.game for selection in pool.square_games]
    if not games and pool.squares_game is not None:
        games = [pool.squares_game]
    return sorted(games, key=lambda game: (game.start_time, game.game_id))


def _lock_time(pool: models.Pool) -> datetime:
    games = _selected_games(pool)
    if not games:
        raise HTTPException(status_code=409, detail="This Squares board has no selected games")
    return games[0].start_time


def _is_admin(db: Session, pool: models.Pool, user: models.User) -> bool:
    return bool(
        is_platform_super_admin(user)
        or pool.owner_id == user.id
        or db.query(models.PoolAdmin).filter(
            models.PoolAdmin.pool_id == pool.id,
            models.PoolAdmin.user_id == user.id,
        ).first()
    )


def _require_participant(db: Session, pool: models.Pool, user: models.User) -> None:
    if not is_platform_super_admin(user) and not is_pool_participant(db, pool.id, user.id):
        raise HTTPException(status_code=403, detail="Pool membership required")


def _lock_board(board: models.SquareBoard, *, actor_id: str | None = None) -> bool:
    if board.locked_at:
        return False
    generator = secrets.SystemRandom()
    board.home_digits = json.dumps(generator.sample(range(10), 10), separators=(",", ":"))
    board.away_digits = json.dumps(generator.sample(range(10), 10), separators=(",", ":"))
    board.locked_at = _now()
    board.locked_by = actor_id
    board.updated_at = board.locked_at
    return True


def ensure_board_locked_for_kickoff(db: Session, pool: models.Pool) -> bool:
    """Automatically lock at kickoff so a missing commissioner action cannot expose digits."""
    if pool.square_board.locked_at or _now() < _lock_time(pool):
        return False
    db.query(models.SquareBoard).filter(models.SquareBoard.pool_id == pool.id).with_for_update().one()
    db.refresh(pool.square_board)
    changed = _lock_board(pool.square_board)
    if changed:
        db.commit()
    return changed


def _claim_payload(claim: models.SquareClaim) -> dict:
    return {
        "id": claim.id,
        "row_index": claim.row_index,
        "column_index": claim.column_index,
        "block_number": claim.row_index * 10 + claim.column_index + 1,
        "user_id": claim.user_id,
        "user_email": claim.user.email,
        "display_name": claim.display_name,
        "claimed_at": claim.claimed_at,
    }


def _effective_total_pot(board: models.SquareBoard, claim_count: int) -> int | None:
    if board.pot_mode == "per_square":
        return board.per_square_cents * claim_count if board.per_square_cents is not None else None
    return board.total_pot_cents


def board_payload(db: Session, pool: models.Pool, user: models.User) -> dict:
    ensure_board_locked_for_kickoff(db, pool)
    db.refresh(pool.square_board)
    board = pool.square_board
    claims = db.query(models.SquareClaim).options(joinedload(models.SquareClaim.user)).filter(
        models.SquareClaim.pool_id == pool.id
    ).all()
    claims_by_cell = {(claim.row_index, claim.column_index): claim for claim in claims}
    payouts = db.query(models.SquarePayout).options(joinedload(models.SquarePayout.winner)).filter(
        models.SquarePayout.pool_id == pool.id
    ).all()
    games = _selected_games(pool)
    if not games:
        raise HTTPException(status_code=409, detail="This Squares board has no selected games")
    game_by_id = {game.game_id: game for game in games}
    primary_game = games[0]
    admin = _is_admin(db, pool, user)
    used_capacity, capacity_limit, plan = entitlements.capacity_usage(db, pool)
    commissioner_controls = plan not in ("free", "squares-plus")
    members = []
    if admin:
        members = [{"id": member.id, "email": member.email} for member in (
            db.query(models.User).join(models.PoolMember, models.PoolMember.user_id == models.User.id)
            .filter(models.PoolMember.pool_id == pool.id).order_by(models.User.email).all()
        )]
    return {
        "pool_id": pool.id,
        "pool_name": pool.name,
        "game": {
            "game_id": primary_game.game_id,
            "start_time": primary_game.start_time,
            "status": primary_game.status,
            "home_team": {"id": primary_game.home_team_id, "name": primary_game.home_team.name, "abbrv": primary_game.home_team.abbrv},
            "away_team": {"id": primary_game.away_team_id, "name": primary_game.away_team.name, "abbrv": primary_game.away_team.abbrv},
            "home_score": primary_game.home_score,
            "away_score": primary_game.away_score,
        },
        "games": [{
            "game_id": game.game_id,
            "start_time": game.start_time,
            "status": game.status,
            "home_team": {"id": game.home_team_id, "name": game.home_team.name, "abbrv": game.home_team.abbrv},
            "away_team": {"id": game.away_team_id, "name": game.away_team.name, "abbrv": game.away_team.abbrv},
            "home_score": game.home_score,
            "away_score": game.away_score,
        } for game in games],
        "lock_time": games[0].start_time,
        "locked": board.locked_at is not None,
        "locked_at": board.locked_at,
        "home_digits": json.loads(board.home_digits) if board.home_digits else None,
        "away_digits": json.loads(board.away_digits) if board.away_digits else None,
        "pot_mode": board.pot_mode or "fixed",
        "total_pot_cents": _effective_total_pot(board, len(claims)),
        "per_square_cents": board.per_square_cents,
        "payout_percentages": {"q1": board.q1_percent, "halftime": board.halftime_percent, "q3": board.q3_percent, "final": board.final_percent},
        "claims": [_claim_payload(claim) for claim in claims],
        "payouts": [{
            "game_id": payout.game_id,
            "game": {
                "home_team": game_by_id[payout.game_id].home_team.abbrv,
                "away_team": game_by_id[payout.game_id].away_team.abbrv,
            } if payout.game_id in game_by_id else None,
            "checkpoint": payout.checkpoint,
            "home_score": payout.home_score,
            "away_score": payout.away_score,
            "winning_row": payout.winning_row,
            "winning_column": payout.winning_column,
            "winner_user_id": payout.winner_user_id,
            "winner_email": payout.winner.email if payout.winner else None,
            "winner_display_name": (
                claims_by_cell[(payout.winning_row, payout.winning_column)].display_name
                if (payout.winning_row, payout.winning_column) in claims_by_cell
                else None
            ),
            "amount_cents": payout.amount_cents,
            "determined_at": payout.determined_at,
        } for payout in payouts],
        "members": members,
        "plan": plan,
        "block_limit": min(capacity_limit, 100) if capacity_limit is not None else 100,
        "permissions": {
            "is_admin": admin,
            "can_claim": board.locked_at is None and _now() < games[0].start_time and (capacity_limit is None or used_capacity < capacity_limit),
            "can_admin_assign": admin and commissioner_controls,
            "can_use_variable_pot": admin and commissioner_controls,
        },
    }


@router.get("/{pool_id}")
def get_board(pool_id: str, db: Session = Depends(deps.get_db), current_user: models.User = Depends(deps.get_current_user)):
    pool = _pool(db, pool_id)
    _require_participant(db, pool, current_user)
    return board_payload(db, pool, current_user)


@router.post("/{pool_id}/claims", status_code=status.HTTP_201_CREATED)
def claim_square(pool_id: str, request: schemas.SquareClaimCreate, db: Session = Depends(deps.get_db), current_user: models.User = Depends(deps.get_current_user)):
    pool = _pool(db, pool_id)
    _require_participant(db, pool, current_user)
    admin = _is_admin(db, pool, current_user)
    target_id = request.user_id or current_user.id
    if target_id != current_user.id and not admin:
        raise HTTPException(status_code=403, detail="Only a pool admin may assign a square to another member")
    if target_id != current_user.id and entitlements.pool_plan(db, pool) in ("free", "squares-plus"):
        raise HTTPException(status_code=403, detail="Upgrade to Commish to assign blocks for other members")
    if not is_pool_participant(db, pool.id, target_id):
        raise HTTPException(status_code=400, detail="Squares may only be assigned to pool members")
    db.query(models.SquareBoard).filter(models.SquareBoard.pool_id == pool.id).with_for_update().one()
    db.refresh(pool.square_board)
    if pool.square_board.locked_at or _now() >= _lock_time(pool):
        raise HTTPException(status_code=409, detail="This Squares board is locked")
    entitlements.enforce_entry_capacity(db, pool)
    claim = models.SquareClaim(
        id=str(uuid.uuid4()), pool_id=pool.id, row_index=request.row_index,
        column_index=request.column_index, user_id=target_id, assigned_by=current_user.id,
        display_name=request.display_name, claimed_at=_now(),
    )
    db.add(claim)
    try:
        db.commit()
    except IntegrityError:
        db.rollback()
        raise HTTPException(status_code=409, detail="That square has already been claimed")
    block_number = claim.row_index * 10 + claim.column_index + 1
    create_audit_log(db, "CLAIM_SQUARE", f"Reserved block {block_number}", current_user.id, "square", claim.id, {"pool_id": pool.id, "assigned_to": target_id, "display_name": claim.display_name, "block_number": block_number})
    return {"id": claim.id, "row_index": claim.row_index, "column_index": claim.column_index, "block_number": block_number, "user_id": claim.user_id, "display_name": claim.display_name}


@router.delete("/{pool_id}/claims/{claim_id}", status_code=status.HTTP_204_NO_CONTENT)
def release_square(pool_id: str, claim_id: str, db: Session = Depends(deps.get_db), current_user: models.User = Depends(deps.get_current_user)):
    pool = _pool(db, pool_id)
    _require_participant(db, pool, current_user)
    claim = db.query(models.SquareClaim).filter(models.SquareClaim.id == claim_id, models.SquareClaim.pool_id == pool.id).first()
    if not claim:
        raise HTTPException(status_code=404, detail="Square claim not found")
    if claim.user_id != current_user.id and not _is_admin(db, pool, current_user):
        raise HTTPException(status_code=403, detail="You cannot release another member's square")
    if pool.square_board.locked_at or _now() >= _lock_time(pool):
        raise HTTPException(status_code=409, detail="This Squares board is locked")
    block_number = claim.row_index * 10 + claim.column_index + 1
    db.delete(claim)
    db.commit()
    create_audit_log(db, "RELEASE_SQUARE", f"Released block {block_number}", current_user.id, "square", claim_id, {"pool_id": pool.id, "block_number": block_number})
    return Response(status_code=status.HTTP_204_NO_CONTENT)


@router.patch("/{pool_id}/claims/display-name")
def update_claim_display_name(
    pool_id: str,
    request: schemas.SquareDisplayNameUpdate,
    db: Session = Depends(deps.get_db),
    current_user: models.User = Depends(deps.get_current_user),
):
    """Update the board-facing name on every claim owned by one member."""
    pool = _pool(db, pool_id)
    if not _is_admin(db, pool, current_user):
        raise HTTPException(status_code=403, detail="Pool admin access required")
    claims = db.query(models.SquareClaim).filter(
        models.SquareClaim.pool_id == pool.id,
        models.SquareClaim.user_id == request.user_id,
    ).all()
    if not claims:
        raise HTTPException(status_code=404, detail="No Squares reservations found for this member")
    previous_names = sorted({claim.display_name for claim in claims if claim.display_name})
    for claim in claims:
        claim.display_name = request.display_name
    db.commit()
    create_audit_log(
        db,
        "UPDATE_SQUARE_DISPLAY_NAME",
        f"Updated display name on {len(claims)} Squares reservation(s)",
        current_user.id,
        "pool_user",
        request.user_id,
        {
            "pool_id": pool.id,
            "user_id": request.user_id,
            "previous_display_names": previous_names,
            "display_name": request.display_name,
            "claim_count": len(claims),
        },
    )
    return board_payload(db, pool, current_user)


@router.post("/{pool_id}/lock")
def lock_board(pool_id: str, db: Session = Depends(deps.get_db), current_user: models.User = Depends(deps.get_current_user)):
    pool = _pool(db, pool_id)
    if not _is_admin(db, pool, current_user):
        raise HTTPException(status_code=403, detail="Pool admin access required")
    db.query(models.SquareBoard).filter(models.SquareBoard.pool_id == pool.id).with_for_update().one()
    db.refresh(pool.square_board)
    if pool.square_board.locked_at:
        raise HTTPException(status_code=409, detail="This Squares board is already locked")
    _lock_board(pool.square_board, actor_id=current_user.id)
    db.commit()
    create_audit_log(db, "LOCK_SQUARE_BOARD", "Randomized digits and locked Squares board", current_user.id, "pool", pool.id)
    return board_payload(db, pool, current_user)


@router.patch("/{pool_id}/payouts")
def update_payouts(pool_id: str, request: schemas.SquarePayoutConfig, db: Session = Depends(deps.get_db), current_user: models.User = Depends(deps.get_current_user)):
    pool = _pool(db, pool_id)
    if not _is_admin(db, pool, current_user):
        raise HTTPException(status_code=403, detail="Pool admin access required")
    if pool.square_board.locked_at or _now() >= _lock_time(pool):
        raise HTTPException(status_code=409, detail="Payout settings cannot change after the board locks")
    percentages = [request.q1_percent, request.halftime_percent, request.q3_percent, request.final_percent]
    if sum(percentages) != 100:
        raise HTTPException(status_code=400, detail="Payout percentages must total 100")
    if request.pot_mode == "per_square" and request.per_square_cents is None:
        raise HTTPException(status_code=400, detail="An amount per reserved block is required")
    if request.pot_mode == "per_square" and entitlements.pool_plan(db, pool) in ("free", "squares-plus"):
        raise HTTPException(status_code=403, detail="Upgrade to Commish to calculate the pot per reserved block")
    board = pool.square_board
    board.pot_mode = request.pot_mode
    board.total_pot_cents = request.total_pot_cents if request.pot_mode == "fixed" else None
    board.per_square_cents = request.per_square_cents if request.pot_mode == "per_square" else None
    board.q1_percent, board.halftime_percent, board.q3_percent, board.final_percent = percentages
    board.updated_at = _now()
    db.commit()
    create_audit_log(db, "UPDATE_SQUARE_PAYOUTS", "Updated Squares payout configuration", current_user.id, "pool", pool.id, request.model_dump())
    return board_payload(db, pool, current_user)
